### PR TITLE
Added generated loot tables for living entities

### LINKED ISSUE #379: Entity loot drops appear not to happen after 1.12
## Issue Description
When killing mobs in Tropicraft, loot isn't droping.

## Steps to Reproduce
1) Kill an Iguana
2) Notice no drops
3) Use console commands to give Iguana Leather to make Scuba Harness

Tropicraft Version:   8.1.0
Forge Version:   all versions

## Crash log
No Crash

## Suggested fix
As the goal is to move away from hardcoding drops in the entity files, look into creating the appropriate loot drops via the more modern mechanism that would deliver the same drops as the 1.12 hardcoded `dropLoot` and/or `dropSpecialtems` would.

### PR DESCRIPTION
**Pull request by ramou**
* Fixed #379
  * iguana's drop leather and scales
  * frogs (all, not just green, sadly) dro poison frog skin
  * dolphins and sharks drop Tropical Fertilizer
  * Starfish and turtles drop their respective shells
  * Marlins drop fresh marlin
  * River Sardines, Tropical Fish and Piranahs drop raw fish
  * Sea Urchin drop Roe (no measure for age)
  * Eagle Rays drop Raw Eagle Ray
  * Man'o'War drops slime balls

### CODE REVIEW

File: src/main/java/net/tropicraft/core/common/data/TropicraftLootTableProvider.java [hunk 1]
```diff
@@ -63,6 +83,103 @@ protected void validate(Map<ResourceLocation, LootTable> map, ValidationTracker
...
+			this.registerLootTable(entity.get(), LootTable.builder());
+		}
+		
+        @Override
+        protected Iterable<EntityType<?>> getKnownEntities() {
+        	return TropicraftEntities.ENTITIES.getEntries().stream().map(Supplier::get).collect(Collectors.toList());
+        }
+        
+        
+    }
```
Comment (tterrag1098) [inline src/main/java/net/tropicraft/core/common/data/TropicraftLootTableProvider.java]: Inconsistent indentation all over here

File: src/main/java/net/tropicraft/core/common/data/TropicraftLootTableProvider.java [hunk 2]
```diff
@@ -63,6 +83,103 @@ protected void validate(Map<ResourceLocation, LootTable> map, ValidationTracker
...
+			registerEntityLoot(TropicraftEntities.MAN_O_WAR, Items.SLIME_BALL, RandomValueRange.of(3, 4));
+}
+
+		//Only Green frogs should drop poison skins!
+    	private static final ILootCondition.IBuilder GREEN = 
+    			EntityHasProperty.builder(LootContext.EntityTarget.THIS, 
+    					EntityPredicate.Builder.create().flags(
+    							new EntityFlagsPredicate(null, null, null, null, false)).nbt(
+    									new NBTPredicate(Util.make(new CompoundNBT(), c -> c.putInt("Type", TreeFrogEntity.Type.GREEN.ordinal())))));
+    	public void registerFrogLoot() {
```
Comment (tterrag1098) [inline src/main/java/net/tropicraft/core/common/data/TropicraftLootTableProvider.java]: This method should be made more reusable or just inlined into the above code, not much point in defining a method that can only be used once.

File: src/main/java/net/tropicraft/core/common/data/TropicraftLootTableProvider.java [hunk 3]
```diff
@@ -63,6 +83,103 @@ protected void validate(Map<ResourceLocation, LootTable> map, ValidationTracker
...
+			registerMinimalLootTable(TropicraftEntities.COWKTAIL);
+			registerEntityLoot(TropicraftEntities.MAN_O_WAR, Items.SLIME_BALL, RandomValueRange.of(3, 4));
+}
+
+		//Only Green frogs should drop poison skins!
+    	private static final ILootCondition.IBuilder GREEN = 
+    			EntityHasProperty.builder(LootContext.EntityTarget.THIS, 
+    					EntityPredicate.Builder.create().flags(
+    							new EntityFlagsPredicate(null, null, null, null, false)).nbt(
+    									new NBTPredicate(Util.make(new CompoundNBT(), c -> c.putInt("Type", TreeFrogEntity.Type.GREEN.ordinal())))));
```
Comment (tterrag1098) [inline src/main/java/net/tropicraft/core/common/data/TropicraftLootTableProvider.java]: The indentation nesting here is a bit excessive
```suggestion
    	private static final ILootCondition.IBuilder GREEN = 
    			EntityHasProperty.builder(LootContext.EntityTarget.THIS, EntityPredicate.Builder.create()
    			    .flags(new EntityFlagsPredicate(null, null, null, null, false))
    			    .nbt(new NBTPredicate(Util.make(new CompoundNBT(), c -> c.putInt("Type", TreeFrogEntity.Type.GREEN.ordinal())))));
```

File: src/main/java/net/tropicraft/core/common/data/TropicraftLootTableProvider.java [hunk 4]
```diff
@@ -63,6 +83,103 @@ protected void validate(Map<ResourceLocation, LootTable> map, ValidationTracker
...
+    	//Drops several items that are affected by Enchantment
+    	//Looting will at most double yield with Looting III
+		public <T extends LivingEntity> void registerEntityLoot(RegistryObject<EntityType<T>> entity, Item loot, IRandomRange range) {
+			this.registerLootTable(entity.get(), LootTable.builder()
+					.addLootPool(LootPool.builder().rolls(range)
+							.addEntry(ItemLootEntry.builder(loot)
+									.acceptFunction(LootingEnchantBonus.builder(RandomValueRange.of(0F, 1F / 3F))))));
+		}
+		
+		public <T extends LivingEntity> void registerMinimalLootTable(RegistryObject<EntityType<T>> entity) {
```
Comment (tterrag1098) [inline src/main/java/net/tropicraft/core/common/data/TropicraftLootTableProvider.java]: Maybe just call this `noDrops` ?

File: src/main/java/net/tropicraft/core/common/data/TropicraftLootTableProvider.java [hunk 5]
```diff
@@ -63,6 +83,103 @@ protected void validate(Map<ResourceLocation, LootTable> map, ValidationTracker
...
+		}
+
+    	//Drops several items that are affected by Enchantment
+    	//Looting will at most double yield with Looting III
+		public <T extends LivingEntity> void registerEntityLoot(RegistryObject<EntityType<T>> entity, Item loot, IRandomRange range) {
+			this.registerLootTable(entity.get(), LootTable.builder()
+					.addLootPool(LootPool.builder().rolls(range)
+							.addEntry(ItemLootEntry.builder(loot)
+									.acceptFunction(LootingEnchantBonus.builder(RandomValueRange.of(0F, 1F / 3F))))));
+		}
```
Comment (tterrag1098) [inline src/main/java/net/tropicraft/core/common/data/TropicraftLootTableProvider.java]: These methods could use more distinct names, such as `dropItem`, `dropItemWithEnchantBonus`, etc.

File: src/main/java/net/tropicraft/core/common/data/TropicraftLootTableProvider.java [hunk 6]
```diff
@@ -63,6 +83,103 @@ protected void validate(Map<ResourceLocation, LootTable> map, ValidationTracker
...
+    	public void registerFrogLoot() {
+			this.registerLootTable(TropicraftEntities.TREE_FROG.get(), LootTable.builder()
+				.addLootPool(LootPool.builder().rolls(ConstantRange.of(1)).acceptCondition(GREEN)
+					.addEntry(ItemLootEntry.builder(TropicraftItems.POISON_FROG_SKIN.get())))
+				.addLootPool(LootPool.builder().rolls(ConstantRange.of(2))
+							.addEntry(ItemLootEntry.builder(TropicraftItems.FROG_LEG.get())))
+			);
+		}
+
+    	public void registerCreeperLoot() {
```
Comment (tterrag1098) [inline src/main/java/net/tropicraft/core/common/data/TropicraftLootTableProvider.java]: Again, could be inlined or made reusable.

**Review by tterrag1098 — CHANGES_REQUESTED** [review]

**Comment by ramou** [comment]
I'll have to remember how to change eclipse to use spaces instead of tabs, but I think I've made it consistent with the previous code at the bottom. I've incorporated all other suggestions, though didn't come up with anything better for `dropItemsWithEnchantBonus` when the single item isn't affected by enchantments but the multiple ones are..

**Comment by Gegy** [comment]
Merged in https://github.com/Tropicraft/Tropicraft/commit/21cbd80413c5f1f78d53f34bfc166f2ec3bb8aba, tysm for the PR and apologies for the delay!

**Pull request closed without merging**